extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/BeanArchiveProcessor.java

```diff
@@ -75,26 +79,29 @@ public BeanArchiveIndexBuildItem build(LiveReloadBuildItem liveReloadBuildItem)
         }
         Set<DotName> generatedClassNames = new HashSet<>();
         for (GeneratedBeanBuildItem generatedBeanClass : generatedBeans) {
-            IndexingUtil.indexClass(generatedBeanClass.getName(), additionalBeanIndexer, applicationIndex, additionalIndex,
-                    Thread.currentThread().getContextClassLoader(),
-                    generatedBeanClass.getData());
+            IndexingUtil.indexClass(generatedBeanClass.getName(), additionalBeanIndexer, applicationIndex,
+                    additionalIndex, Thread.currentThread().getContextClassLoader(), generatedBeanClass.getData());
             generatedClassNames.add(DotName.createSimple(generatedBeanClass.getName().replace('/', '.')));
-            generatedClass.produce(new GeneratedClassBuildItem(true, generatedBeanClass.getName(), generatedBeanClass.getData(),
-                    generatedBeanClass.getSource()));
+            generatedClass.produce(new GeneratedClassBuildItem(true, generatedBeanClass.getName(),
+                    generatedBeanClass.getData(), generatedBeanClass.getSource()));
         }
 
-        BeanArchives.PersistentClassIndex index = liveReloadBuildItem.getContextObject(BeanArchives.PersistentClassIndex.class);
+        BeanArchives.PersistentClassIndex index = liveReloadBuildItem
+                .getContextObject(BeanArchives.PersistentClassIndex.class);
         if (index == null) {
             index = new BeanArchives.PersistentClassIndex();
             liveReloadBuildItem.setContextObject(BeanArchives.PersistentClassIndex.class, index);
         }
 
         // Finally, index ArC/CDI API built-in classes
-        return new BeanArchiveIndexBuildItem(
-                BeanArchives.buildBeanArchiveIndex(Thread.currentThread().getContextClassLoader(), index, applicationIndex,
-                        additionalBeanIndexer.complete()),
-                generatedClassNames,
-                additionalBeans);
+        BeanArchiveIndexBuildItem i = new BeanArchiveIndexBuildItem(
+                BeanArchives.buildBeanArchiveIndex(Thread.currentThread().getContextClassLoader(), index,
+                        applicationIndex, additionalBeanIndexer.complete()),
+                generatedClassNames, additionalBeans);
+
+        extraIndexBuildItem.produce(new ExtraIndexBuildItem(i.getIndex()));
```

Comment:
I think that you should only extract an index for generated bean classes. `BeanArchiveIndexBuildItem#index` contains a lot more, and even indexes JDK classes on demand.

Reply:
@mkouba please have a look at the updated version, now it uses additionalBeans index instead. Hope this is what you meant.